skip from preline to inline, see issue #63

jhayes45 opened the pull request:
createLine (controllers/lines; POST /lines) now skips users ahead to 'inline' status if there are not yet 5 people in the entire line (line stages 1-3). Returns the updated line, showing that is has 'inline' status.

If there are at least 5 people already in line then it returns the line with 'preline' status, and user has to wait to receive a notification via our cron job (need to debug cron job - #109). 

async/await is used to made the code look and behave synchronously even though it has to wait for/"await" asynchronous db queries to complete.

jhayes45 commented:
note: I tested it both when the batch was not full and full

brianwang9100 commented on src/api/models/lines.js:
```diff
@@ -4,6 +4,8 @@ import idvalidator from 'mongoose-id-validator'
 const Schema = mongoose.Schema;
 const LineEvent = mongoose.model('LineEvent')
 
+const BATCH_SIZE = 5;
```
soo i think theres already a batch size in updateAllPrelineToNotification cron job, can u modify that file to reference this BATCH_SIZE?

brianwang9100 commented on src/api/models/lines.js:
```diff
@@ -43,49 +45,55 @@ lineSchema.plugin(idvalidator);
 
         case 'notification':
             if (this.status !== 'preline') {
-                msg = "LineUpdateError: Cannot move to notification status from status " + this.status;
```
it should be this.status, not status. If youre gonna make an error, u gotta tell them what the original status is. using status will result in “cannot update startrecruiter from status startrecruiter” when the actual status was preline.

jhayes45 replied:
yeah absolutely. good catch

brianwang9100 commented on src/api/models/lines.js:
```diff
@@ -43,49 +45,55 @@ lineSchema.plugin(idvalidator);
         case 'inline':
-            if (this.status !== 'notification') {
-                msg = "LineUpdateError: Cannot move to inline status from status " + this.status;
+            if (this.status !== 'notification' && this.status !== 'preline') {
```
so can u think of a case where a preliner can illegally join inline even if BATCH is full? if not this is fine.

jhayes45 replied:
if they figure out our API then they can

jhayes45 replied:
i'll make an issue for this

jhayes45 replied:
#114

brianwang9100 commented on src/api/models/lines.js:
```diff
@@ -99,31 +107,35 @@ lineSchema.methods.updateStatus = async function(status) {
 
     if (msg !== null) {
-        return msg;
+        return Promise.reject(msg);
```
what functionality does Promise.reject have when u step out of the function? do .catch statements catch rejected promises?

jhayes45 replied:
I added try/catch to updateLineStatus (the patch route) since it lacked it, and also to answer your question.

Here is the code:
![image](https://user-images.githubusercontent.com/35344365/37002132-aa0534d0-2096-11e8-9605-0e09b46ecdcf.png)

And here is the result, trying to send a line from notification to startRecruiter. Obviously, this is disallowed, so the promise rejects, just as the code you highlighted showed. This is caught by the code in the above screenshot, and you can see that the client is returned the message that the promise rejected with:
![image](https://user-images.githubusercontent.com/35344365/37002021-547dd3c8-2096-11e8-92b7-c50e3bfab252.png)

jhayes45 commented:
The commit that I just pushed:

-fixed status to be this.status in the error message
-added try/catch block to async/await
-changed BATCH_SIZE in the cron job to call the model's getBatchSize()

brianwang9100 commented:
looks good, merging...
